csharp/ql/lib/semmle/code/csharp/dataflow/ExternalFlow.qll

```diff
@@ -504,15 +504,26 @@ class UnboundCallable extends Callable {
   }
 }
 
+pragma[nomagic]
+private predicate callableSpecInfo(Callable c, string namespace, string type, string name) {
+  c.getDeclaringType().hasQualifiedName(namespace, type) and
+  c.getName() = name
+}
+
+pragma[nomagic]
+private predicate subtypeSpecCandidate(Callable c, UnboundValueOrRefType t) {
+  elementSpec(_, _, true, c.getName(), _, _, t)
```

Comment:
This looks like a potentially bad join, as we will relate *all* calls with matching names. How about first joining on `getASubTypeUnbound+()` instead?

```ql
pragma[nomagic]
private predicate subtypeSpecCandidate(string name, UnboundValueOrRefType t) {
  exists(UnboundValueOrRefType t0 |
    elementSpec(_, _, true, name, _, _, t0) and
    t = t0.getASubTypeUnbound+()
  )
}

pragma[nomagic]
private predicate callableInfo(Callable c, string name, UnboundValueOrRefType decl) {
  name = c.getName() and
  decl = c.getDeclaringType()
}
```

and then use it in `InterpretedCallable` as

```ql
exists(string name, UnboundValueOrRefType t |
  callableInfo(this, name, t) and
  subtypeSpecCandidate(name, t)
)
```

Reply:
Yes, I think you are right! However, in this particular case the join is worsened slightly.
Let us try with your solution and run DCA.

<img width="1038" alt="image" src="https://user-images.githubusercontent.com/6472811/191809215-d850ae80-3023-4cfb-b104-d3c8c19ea515.png">
<img width="1164" alt="image" src="https://user-images.githubusercontent.com/6472811/191809339-6543061e-0caa-4f6c-8087-c9707c6544a6.png">
<img width="1166" alt="image" src="https://user-images.githubusercontent.com/6472811/191809449-770daae7-921e-41f3-aa57-b0e03813a4f2.png">

Reply:
> However, in this particular case the join is worsened slightly.

I think that is only because we are currently lucky that `elementSpec(_, _, true, name, _, _, t)` is rather small. But even so, if, say, `elementSpec(_, _, "ToString", _, _, t)` held, we would join with *all* `ToString` methods, which can be a large set.

Reply:
Yes - I totally agree.